reference/test/base/reordered.cpp

```diff
@@ -0,0 +1,487 @@
+    auto scaled_reordered = scaled_reordered_fact->generate(this->mtx);
+
+    GKO_ASSERT_MTX_NEAR(gko::as<Mtx>(scaled_reordered->get_system_matrix()),
+                        this->row_scaled, this->tol);
+}
+
+
+TYPED_TEST(ScaledReordered, BuildsWithColScaling)
+{
+    using SR = typename TestFixture::SR;
+    using Mtx = typename TestFixture::Mtx;
+    auto scaled_reordered_fact =
+        SR::build().with_col_scaling(this->diag).on(this->exec);
+    auto scaled_reordered = scaled_reordered_fact->generate(this->mtx);
+
+    GKO_ASSERT_MTX_NEAR(gko::as<Mtx>(scaled_reordered->get_system_matrix()),
+                        this->col_scaled, this->tol);
+}
+
+
+TYPED_TEST(ScaledReordered, BuildsWithRowAndColScaling)
+{
+    using SR = typename TestFixture::SR;
+    using Mtx = typename TestFixture::Mtx;
+    auto scaled_reordered_fact = SR::build()
+                                     .with_row_scaling(this->diag)
+                                     .with_col_scaling(this->diag)
+                                     .on(this->exec);
+    auto scaled_reordered = scaled_reordered_fact->generate(this->mtx);
+
+    GKO_ASSERT_MTX_NEAR(gko::as<Mtx>(scaled_reordered->get_system_matrix()),
+                        this->row_col_scaled, this->tol);
+}
+
+
+TYPED_TEST(ScaledReordered, BuildsWithRcmReordering)
+{
+    using SR = typename TestFixture::SR;
+    using Mtx = typename TestFixture::Mtx;
+    auto scaled_reordered_fact =
+        SR::build().with_reordering(this->rcm_factory).on(this->exec);
+    auto scaled_reordered = scaled_reordered_fact->generate(this->rcm_mtx);
+
+    GKO_ASSERT_MTX_NEAR(gko::as<Mtx>(scaled_reordered->get_system_matrix()),
+                        this->rcm_mtx_reordered, this->tol);
+}
+
+
+TYPED_TEST(ScaledReordered, ThrowOnNonSquareMatrix)
+{
+    using SR = typename TestFixture::SR;
+    auto scaled_reordered_fact = SR::build().on(this->exec);
+
+    ASSERT_THROW(scaled_reordered_fact->generate(this->rectangular_mtx),
+                 gko::DimensionMismatch);
+}
+
+
+TYPED_TEST(ScaledReordered, CanBeCopied)
+{
+    using SR = typename TestFixture::SR;
+    auto scaled_reordered_fact = SR::build()
+                                     .with_row_scaling(this->diag)
+                                     .with_col_scaling(this->diag)
+                                     .with_reordering(this->rcm_factory)
+                                     .with_inner_operator(this->solver_factory)
+                                     .on(this->exec);
+    auto scaled_reordered = scaled_reordered_fact->generate(this->mtx);
+
+    auto before_system_matrix = scaled_reordered->get_system_matrix();
+    auto before_inner_operator = scaled_reordered->get_inner_operator();
+
+    auto copied = SR::build().on(this->exec)->generate(this->rcm_mtx);
+    copied->copy_from(scaled_reordered.get());
+
+    ASSERT_EQ(before_system_matrix, copied->get_system_matrix());
+    ASSERT_EQ(before_inner_operator, copied->get_inner_operator());
```

Comment:
nit: for cleaner AAA separation:
```suggestion
    using SR = typename TestFixture::SR;
    auto scaled_reordered_fact = SR::build()
                                     .with_row_scaling(this->diag)
                                     .with_col_scaling(this->diag)
                                     .with_reordering(this->rcm_factory)
                                     .with_inner_operator(this->solver_factory)
                                     .on(this->exec);
    auto scaled_reordered = scaled_reordered_fact->generate(this->mtx);
    auto before_system_matrix = scaled_reordered->get_system_matrix();
    auto before_inner_operator = scaled_reordered->get_inner_operator();

    auto copied = SR::build().on(this->exec)->generate(this->rcm_mtx);
    copied->copy_from(scaled_reordered.get());

    ASSERT_EQ(before_system_matrix, copied->get_system_matrix());
    ASSERT_EQ(before_inner_operator, copied->get_inner_operator());
```